### PR TITLE
Making changes for collision paper (widgets and bugs)

### PR DESCRIPTION
**Pull request by yger**

### CODE REVIEW

File: spikeinterface/comparison/collisioncomparison.py [hunk 1]
```diff
@@ -42,8 +42,8 @@ def get_label_for_collision(self, gt_unit_id1, gt_unit_id2):
         mask = (self.collision_events['unit_id1'] == gt_unit_id1) & (self.collision_events['unit_id2'] == gt_unit_id2)
         event = self.collision_events[mask]
         
-        score_label1 = self._labels_st1[gt_unit_id1][event['index1']]
-        score_label2 = self._labels_st1[gt_unit_id2][event['index2']]
+        score_label1 = self._labels_st1[self.unit1_ids[gt_index1]][event['index1']]
+        score_label2 = self._labels_st1[self.unit1_ids[gt_index2]][event['index2']]
```
Comment (samuelgarcia) [inline spikeinterface/comparison/collisioncomparison.py:46]: More difficult to read no ?

File: spikeinterface/widgets/correlograms.py [hunk 2]
```diff
@@ -44,16 +43,18 @@ def plot(self):
         for i, unit_id1 in enumerate(unit_ids):
             for j, unit_id2 in enumerate(unit_ids):
                 ccg = correlograms[i, j]
-                ax = self.axes[i, j]
+                ax = self.get_tiled_ax(i+j*len(unit_ids), len(unit_ids), len(unit_ids))
+                #ax = self.axes[i, j]
```
Comment (samuelgarcia) [inline spikeinterface/widgets/correlograms.py:47]: This will break some code I think.

File: spikeinterface/widgets/unitwaveforms.py [hunk 3]
```diff
@@ -139,6 +139,7 @@ def _do_plot(self):
                 if self._plot_waveforms and self._plot_templates:
                     color = 'k'
                 ax.plot(xvectors_flat, template.T.flatten(), lw=1, color=color)
+                ax.set_title('template %s' %unit_ids[i])
```
Comment (samuelgarcia) [inline spikeinterface/widgets/unitwaveforms.py:142]: Use `'f{}'` style formatting.

Comment (yger) [reply spikeinterface/widgets/unitwaveforms.py:142]: Ok, will do

File: spikeinterface/widgets/correlograms.py (hunk 2 above)
Comment (yger) [reply spikeinterface/widgets/correlograms.py:47]: Maybe but it had to be consistent with plot_waveforms and allow plots in ax axis

File: spikeinterface/comparison/collisioncomparison.py (hunk 1 above)
Comment (yger) [reply spikeinterface/comparison/collisioncomparison.py:46]: Yes, I'll revert that. I was just reading everything to be sure I understood it properly.

File: spikeinterface/widgets/unitwaveforms.py (hunk 3 above)
Comment (samuelgarcia) [reply spikeinterface/widgets/unitwaveforms.py:142]: I think I will refactor BaseWIdget and MultiWidget into the same object to get something more convinient in any case.
When ready you will be able to rebase on top of it.

**Comment by yger** [comment]
Can it be possible for a study object to retrieve templates for all sorters? There is the get_metrics() function, but it would be good to more generally have a get_templates(rec_name, sorter_name). Some caching would be done, maybe, and templates estimated? Even better, would also need to work for ground truth if we want to do comparisons

**Comment by alejoe91** [comment]
> Can it be possible for a study object to retrieve templates for all sorters? There is the get_metrics() function, but it would be good to more generally have a get_templates(rec_name, sorter_name). Some caching would be done, maybe, and templates estimated? Even better, would also need to work for ground truth if we want to do comparisons

@yger I think this is a great idea! The study could wrap postprocessing functions. @samuelgarcia what do you think?

**Comment by yger** [comment]
I made a prototype here in my branch. However, not sure about how to optimize everything, since currently, this is always recomputed while caching could be made with your Waveforms folders/objects. Don't know exactly what you have in mind. But this is working :-)

**Comment by yger** [comment]
I'm also assuming that you guys want to work as much as possible with pandas dataframes objects, instead of dictionary. I'm not familiar with them, so for now my widgets are dealing with dict... But this should be easy to change.

**Comment by samuelgarcia** [comment]
Hi Pierre,
I don't have time straight no to check your PR.
I think it is easy to implement to run waveform_extractor for all (sorter, recording) pair and cache it inside the study folder.
Then making a merthod to retrieve template will be easy.

The difficult part is that recording are not nercesarly  pre-processed (filter+whitten+cmr) in a study but for waveform extarctor we need a decide on a stanard pre processing chain at the study level to compute this.
I don't remember how I did this for SNR metric maybe only filter is enought for metrics.

I propose to have a call for this.

**Comment by yger** [comment]
Call whenever you want. Indeed, the method I've implemented does not assume anything for the recording. But since in my case I just want the templates to compute similarities, this does not matter much.

File: spikeinterface/comparison/groundtruthstudy.py [hunk 4]
```diff
@@ -54,22 +54,18 @@ def create(cls, study_folder, gt_dict, **job_kwargs):
         setup_comparison_study(study_folder, gt_dict,  **job_kwargs)
         return cls(study_folder)
 
-    def run_sorters(self, sorter_list, mode_if_folder_exists='keep', **kargs):
+    def run_sorters(self, sorter_list, **kargs):
```
Comment (samuelgarcia) [inline spikeinterface/comparison/groundtruthstudy.py:57]: Non.
Je ne veux la même valeur par défault que la fonction `si.sorters.run_sorter()`.
Pour les study keep c'est mieux, je pense.

File: spikeinterface/comparison/groundtruthstudy.py [hunk 5]
```diff
@@ -203,28 +199,39 @@ def aggregate_dataframes(self, copy_into_folder=True, **karg_thresh):
...
         rec = self.get_recording(rec_name)
-        gt_sorting = self.get_ground_truth(rec_name)
-        
+        if sorter_name is None:
+            sorting = self.get_ground_truth(rec_name)
+            sorter_name = 'gt'
+        else:
+            assert sorter_name in self.sorter_names
+            sorting = self.get_sorting(sorter_name, rec_name)
+
```
Comment (samuelgarcia) [inline spikeinterface/comparison/groundtruthstudy.py:213]: Here I would prefer something super explicit.
This sorter_name=None is not enought explicit.
For me at the study level metrics are GT based.
So if it make sens in some situation to have metrics on sorting ouput I would prefer an other method for this.
Even if internally the implemtation is like this.

File: spikeinterface/comparison/groundtruthstudy.py [hunk 6]
```diff
@@ -203,28 +199,39 @@ def aggregate_dataframes(self, copy_into_folder=True, **karg_thresh):
...
-        we = WaveformExtractor.create(rec, gt_sorting, waveform_folder)
+        we = WaveformExtractor.create(rec, sorting, waveform_folder)
         we.set_params(ms_before=ms_before, ms_after=ms_after, max_spikes_per_unit=max_spikes_per_unit)
         we.run(n_jobs=n_jobs, total_memory=total_memory)
         
         # metrics
-        metrics = compute_quality_metrics(we, metric_names=metric_names)
-        filename = self.study_folder / 'metrics' / f'metrics _{rec_name}.txt'
-        metrics.to_csv(filename, sep='\t', index=True)
+        if templates_only:
```
Comment (samuelgarcia) [inline spikeinterface/comparison/groundtruthstudy.py:224]: don't make sens with this method name.
Maybe you can create 2 methods one for waveform and one for metrics.

File: spikeinterface/comparison/groundtruthstudy.py [hunk 7]
```diff
@@ -203,28 +199,39 @@ def aggregate_dataframes(self, copy_into_folder=True, **karg_thresh):
...
+            filename = self.study_folder / 'metrics' / f'templates_{sorter_name}_{rec_name}.npy'
+            metrics = we.get_all_templates()
+            np.save(filename, metrics)
```
Comment (samuelgarcia) [inline spikeinterface/comparison/groundtruthstudy.py:227]: Please do this in separate method.

Comment (samuelgarcia) [reply spikeinterface/comparison/groundtruthstudy.py:227]: C'est pas spyking-cicus ici au niveau orga :)

File: spikeinterface/widgets/connectivitycomp.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        self.ax[1, 0].set_title('auto corr')
+
+        x1 = self.get_traces('true', 'cross')
+        x2 = self.get_traces('estimated', 'cross')
+        self.ax[1, 1].plot(np.abs(x2 - x1).T, '0.5')
+        self.ax[1, 1].plot(np.mean(np.abs(x2 - x1), 0), 'r', lw=2)
+        self.ax[1, 1].set_ylabel('CC')
+        self.ax[1, 1].set_xlabel('time')
+        self.ax[1, 1].set_title('cross corr')
+
```
Comment (samuelgarcia) [inline spikeinterface/widgets/connectivitycomp.py:116]: Please add function on top of the class. See other widgets for that.

File: spikeinterface/widgets/connectivitycomp.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
+import numpy as np
+from matplotlib import pyplot as plt
+
+from .basewidget import BaseWidget
+from spikeinterface.toolkit import compute_correlograms
+
+
+
+class ConnectivityComparisonWidget(BaseWidget):
```
Comment (samuelgarcia) [inline spikeinterface/widgets/connectivitycomp.py:9]: No sure if the name is self describing. More ideas even if it is longer ?

**Comment by samuelgarcia** [comment]
Pierre you should rebase this branch with the current master. It contains more test and bug fix.

**Comment by samuelgarcia** [comment]
close replace by #234

**Pull request closed without merging**